Readme.md
# o:XML
[![Java CI](https://github.com/SamuraiCrow/oXML/actions/workflows/ant.yml/badge.svg)](https://github.com/SamuraiCrow/oXML/actions/workflows/ant.yml)

This is version 1.1.7 of o:XML.  Originally built with JDK 1.3.

You may encounter bugs in this release.  If you do, please report
them; your bug reports are valuable contributions.

The project was hosted by pingdynasty, but now hosted on GitHub because
the source was only salvagable from the Wayback Machine.  The runtime
libraries are already lost.

o:XML is a Java application, and requires a Java runtime environment.
For more information about Java, please see http://java.com

The licence for this software is, unless otherwise stated in the source 
files, the Gnu GPL, which is available in the file licence.txt.

# Future Goals
Recode the backend to generate some other language than Java3.


changes.txt
----- Release 1.1.7
Mar 11 2007 - added support for Spring schema-based configuration
----- Release 1.1.6
Feb 21 2007 - changed db:result parameters to use SQL type mappings
Feb 08 2007 - added Spring 2.0 scripting support
Feb 05 2007 - added functionality to provide alternative SQL commands for different SQL dialects
Jan 24 2007 - updated o:Path parser to accept div/mod/and/or as XML names
Jul 07 2006 - added List.addAll(Nodeset) and List(Nodeset)
Jun 27 2006 - updated o:for-each to maintain the context position during iteration
Jun 10 2006 - added generated nodes() function for Element subtypes
Jun 10 2006 - updated Ant task to process SML files when sml=true
Jun 10 2006 - updated Ant task to load default extensions unless defaultExtensions=false
Jun 10 2006 - updated Ant task with processing mode: run, test or compile
Jun 10 2006 - updated command line tool to process SML unit tests
Jun 08 2006 - changed Java extensions so that Node is a subtype of Object
May 28 2006 - updated to use the SML DocumentBuilder and o:XML macros
May 28 2006 - added -xml command line option
May 21 2006 - resolved problem with faulty instantiation of parent types by copy constructors
May 18 2006 - fixed bug in XPath attribute iterator when iterating over Element subtypes
May 11 2006 - updated static db:function to honour the input parameter type
----- Release 1.1.5
Apr 25 2006 - added -sml command line option
Apr 24 2006 - added -compile command line option
Apr 21 2006 - added -xmldocs and -htmldocs command line options
Apr 20 2006 - fixed a bug to prevent whitespace suppression in o:text evaluation
Apr 04 2006 - added concat(Nodeset) : String
Apr 04 2006 - fixed db:quote() escaped characters (bug #44)
Mar 26 2006 - added Stack type
Mar 26 2006 - added List.add(pos, obj)
Mar 24 2006 - added function narrow(Node typename, Node object)
Mar 24 2006 - added function Node.super(Name typename)
Mar 21 2006 - added o:constructor mapping as synonym for o:function constructor
Mar 21 2006 - added db:function mapping for definition of database functions
Mar 21 2006 - added -test command line option for integrated unit testing
Mar 04 2006 - updated o:for-each to fix problem with iteration over single nodes (bug #43)
Mar 02 2006 - added Servlet URL parameters, and configurable parameter settings
Mar 02 2006 - changed Servlet parameter names: upload-directory, loglevel
Feb 28 2006 - updated namespace declaration output handling (bug #42)
Feb 02 2006 - added function unique(Nodeset) : Nodeset
----- Release 1.1.4
Jan 31 2006 - added compilation check for duplicate type declarations
Jan 30 2006 - added function ends-with(String, String) : Boolean
Jan 30 2006 - added function String.charCodeAt(Number) : Number
Jan 30 2006 - fixed output of default namespace declarations (bug #42)
Sep 26 2005 - fixed Program type parameter scoping (bug #38)
Sep 26 2005 - changed generated functions to be static
Sep 15 2005 - added (output) encoding attribute to ObjectBox Ant task
Sep 13 2005 - added function format-number(num, pattern)
Sep 13 2005 - added function Number.format(pattern)
Sep 13 2005 - added function List.clear()
----- Release 1.1.3
Sep 01 2005 - fixed user-defined function memory leak (bug #37)
----- Release 1.1.2
Apr 30 2005 - fixed threadsafety problem by removing program parameter values from InterpretedProgram
Mar 31 2005 - fixed bug in ObjectBoxService where uploadDirectory is not set
----- Release 1.1.1
Feb 04 2005 - ObjectBoxServlet catches FileNotFoundException and sends 404 - Not Found
Feb 04 2005 - org.oXML.extras.http.ObjectBoxServlet replaces org.oXML.extras.http.ObjectBox
Feb 04 2005 - updated ObjectBoxTask (Ant task) to produce no output when content-type=none, and when 'in' is specified without 'out'
Feb 03 2005 - added content-type="none" as possible o:program attribute
Feb 03 2005 - added Servlet Filter class: org.oXML.extras.http.ObjectBoxFilter
Feb 02 2005 - fixed bug in XPath predicate expression evaluation
Feb 01 2005 - added Servlet 'logLevel' init-param
Feb 01 2005 - updated ObjectBox Servlet to catch program return values: servlet:Redirect, servlet:Forward, servlet:Error, net:Message
Feb 01 2005 - updated ObjectBox CLI to catch and output program return values
----- Release 1.1.0
Nov 30 2004 - attribute valute expression parser updated to allow nested, doubled curly braces and quoted expressions
Nov 18 2004 - StreamResultHandler outputs XML declaration only if content-type is text/xml
Nov 18 2004 - merge result text nodes, as in XSLT
Nov 16 2004 - added CLI -noext and default loading of extensions
Nov 16 2004 - added @value to o:processing-instruction
Nov 16 2004 - added @value to o:attribute
Nov 16 2004 - servlet honours output content-type
Nov 16 2004 - added @content-type to o:program
Nov 15 2004 - added Program type
Nov 14 2004 - added o:meta mapping
Nov 06 2004 - added o:module mapping
Oct 30 2004 - fixed descendant:: and descendant-or-self:: bugs
Oct 24 2004 - added RegularExpression type
Oct 24 2004 - refactored functions to template o:XML constructs
----- Release 1.1.0rc1
Oct 12 2004 - fixed XInclude bug that didn't import XPointer results properly (bug #14)
Oct 10 2004 - supply source file location in exception messages
Oct 09 2004 - refactored db extensions to use full expressions and templates
Oct 06 2004 - updated name and namespace handling to preserve prefixes
Oct 02 2004 - implemented o:session input variable for command-line and servlet environments
Sep 28 2004 - Nodeset refactored to not expose any public type functions
Sep 27 2004 - implemented exceptions with stack traces over o:XML function calls
Sep 25 2004 - refactored core type code templates
Aug 10 2004 - added Array type for multi-dimensional arrays
Aug 10 2004 - updated o:Path expression parser to enable it to produce MLML output
----- Release 1.0.2
Jul 21 2004 - including o:Lib 1.0.0 in the executable jar
Jul 21 2004 - fixed problems with undeclared attribute prefixes and DOM output namespace handling

src/org/oXML/xpath/parser/sym.java

//----------------------------------------------------
// The following code was generated by CUP v0.10j
// Sun Mar 25 21:53:15 CEST 2007
//----------------------------------------------------

package org.oXML.xpath.parser;

/** CUP generated class containing symbol constants. */
public class sym {
  /* terminals */
  public static final int PRECEDING_SIBLING = 39;
  public static final int LTE = 7;
  public static final int ASTERISK = 13;
  public static final int NAMESPACE = 36;
  public static final int EQUALS = 4;
  public static final int SLASH = 23;
  public static final int COMMENT = 41;
  public static final int SLASHSLASH = 24;
  public static final int LPAREN = 9;
  public static final int CHILD = 31;
  public static final int MINUS = 17;
  public static final int PROCESSING_INSTRUCTION = 43;
  public static final int BAR = 15;
  public static final int RPAREN = 10;
  public static final int AND = 21;
  public static final int LT = 5;
  public static final int OR = 20;
  public static final int COMMA = 2;
  public static final int FOLLOWING_SIBLING = 35;
  public static final int DOTDOT = 26;
  public static final int DIV = 18;
  public static final int NC_NAME = 46;
  public static final int DESCENDANT_OR_SELF = 33;
  public static final int PLUS = 16;
  public static final int DOT = 25;
  public static final int DESCENDANT = 32;
  public static final int EOF = 0;
  public static final int error = 1;
  public static final int SELF = 40;
  public static final int NUMBER = 47;
  public static final int MOD = 19;
  public static final int PRECEDING = 38;
  public static final int ANCESTOR_OR_SELF = 29;
  public static final int GTE = 8;
  public static final int ANCESTOR = 28;
  public static final int LITERAL = 45;
  public static final int COLON = 27;
  public static final int LANGLE = 11;
  public static final int NOTEQUALS = 3;
  public static final int DOLLAR = 22;
  public static final int RANGLE = 12;
  public static final int PARENT = 37;
  public static final int FOLLOWING = 34;
  public static final int TEXT = 42;
  public static final int AT = 14;
  public static final int ATTRIBUTE = 30;
  public static final int GT = 6;
  public static final int NODE = 44;
}

